File: app/src/main/java/com/kyhsgeekcode/disassembler/data/RawFile.java
package com.kyhsgeekcode.disassembler.data;

import com.kyhsgeekcode.disassembler.data.AbstractFile;

import java.io.File;

public class RawFile extends AbstractFile {
    public RawFile(File file, byte[] filecontent) {
        this.fileContents = filecontent;
        setPath(file.getAbsolutePath());

    }

}

File: app/src/main/java/com/kyhsgeekcode/disassembler/files/RawFile.kt
package com.kyhsgeekcode.disassembler.files

import java.io.File

class RawFile(
    private val file: File,
    filecontent: ByteArray?,
    private val deferredContentLoader: (() -> ByteArray)? = null
) : AbstractFile() {
    private var contentLoaded = false

    init {
        if (filecontent != null) {
            fileContents = filecontent
            contentLoaded = true
        }
        path = file.path
    }

    override fun getBinaryContents(): ByteArray {
        if (!contentLoaded) {
            fileContents = deferredContentLoader?.invoke()
                ?: throw IllegalStateException("No file contents available for raw binary")
            contentLoaded = true
        }
        return fileContents
    }

    override fun getBinaryLength(): Long {
        return if (contentLoaded) {
            fileContents.size.toLong()
        } else {
            file.length()
        }
    }
}
